sklearn/linear_model/_glm/distribution.py

```diff
@@ -0,0 +1,382 @@
+"""
+Distribution functions used in GLM
+"""
+
+# Author: Christian Lorentzen <lorentzen.ch@googlemail.com>
+# License: BSD 3 clause
+
+from abc import ABCMeta, abstractmethod
+from collections import namedtuple
+import numbers
+
+import numpy as np
+from scipy.special import xlogy
+
+
+def _safe_lin_pred(X, coef):
+    """Compute the linear predictor taking care if intercept is present."""
+    if coef.size == X.shape[1] + 1:
+        return X @ coef[1:] + coef[0]
+    else:
+        return X @ coef
+
+
+DistributionBoundary = namedtuple("DistributionBoundary",
+                                  ("value", "inclusive"))
+
+
+class ExponentialDispersionModel(metaclass=ABCMeta):
+    """Base class for reproductive Exponential Dispersion Models (EDM).
+
+    The pdf of Y∼EDM(μ, φ) is given by::
+
+        p(y| θ, φ) = c1(y, φ) * exp((θy-A(θ))/φ)
+        = c2(y, φ) * exp(-d(y, μ)/(2φ))
+
+    with mean E[Y] = A'(θ) = μ, variance Var[Y] = φ * v(μ),
+    unit variance v(μ), unit deviance d(y,μ) and dispersion parameter φ.
+
+    Methods
+    -------
+    deviance
+    deviance_derivative
+    in_y_range
+    unit_deviance
+    unit_deviance_derivative
+    unit_variance
+    unit_variance_derivative
+
+    References
+    ----------
+    https://en.wikipedia.org/wiki/Exponential_dispersion_model.
+    """
+
+    def in_y_range(self, y):
+        """Returns ``True`` if y is in the valid range of Y∼EDM.
+
+        Parameters
+        ----------
+        y : array, shape (n_samples,)
+            Target values.
+        """
+        if hasattr(self, '_upper_bound'):
+            # All currently supported distributions have an upper bound at
+            # +inf, however this may need to be implemented for other
+            # distributions
+            raise NotImplementedError
+
+        if not isinstance(self._lower_bound, DistributionBoundary):
+            raise TypeError('_lower_bound attribute must be of type '
+                            'DistributionBoundary')
+
+        if self._lower_bound.inclusive:
+            return np.greater_equal(y, self._lower_bound.value)
+        else:
+            return np.greater(y, self._lower_bound.value)
+
+    @abstractmethod
+    def unit_variance(self, mu):
+        """Compute the unit variance function.
+
+        The unit variance v(μ) determines the variance as a function of the
+        mean μ by Var[Y_i] = φ/s_i * v(μ_i).
+        It can also be derived from the unit deviance d(y,μ) as::
+
+            v(μ) = 2/(∂^2 d(y,μ)/(∂ μ^2))|_{y=μ}
```

Comment:
@lorentzenchr Unfortunately we can't use Unicode characters in docstrings. Even if CI passes, it's going to fail on master when it will try to build the PDF of the documentation with LaTeX. So we would need to revert part of 61dc13fdd5. I can do it if you are fine with it.

Reply:
No LaTeX, no unicode 😢  I didn't know that. It's not very important as those are private classes not rendered on the webpage. But what is allowed to use then? Maybe best to let it as it is with LaTeX?

Reply:
>> We don't use latex in the docstrings, only in the user guide.
>
> No LaTeX, no unicode 😢  I didn't know that.

I think it's fine to have some LaTeX. I wasn't aware we had a rule against it. It's certainly a bit less readable in code but having nice formulas in HTML is worth it IMO (better than ascii text in any case). For instance pytorch uses it extensively (e.g. https://pytorch.org/docs/stable/nn.html#torch.nn.L1Loss) and I find their docs quite pretty. Even if it's not part of the public HTML documentation for now, it would be good to have it.  Besides `neighbors.dist_metrics` and tsne has some latex in docstrings already.

We can reduce the amount of Latex in docstrings by moving the notes to the user manual (I think that was another comment).

> No need to use raw strings for docstrings

And if we do use a little bit of Latex, raw docstrings are handy. FWIW it's part of [PEP 257](https://www.python.org/dev/peps/pep-0257/),
```
Use r"""raw triple double quotes""" if you use any backslashes in your docstrings.
```

(I have reverted the commit removing Latex BTW). Will look into reducing the amount of it next.